vendor/purs/runtime/pstring.ss

```diff
@@ -61,33 +62,167 @@
+  ; * When a single continuous buffer is required and the string is a `Concat`,
+  ;   then a new string buffer is allocated and the chunks are copied in linear
+  ;   time to that new buffer. This new buffer is memoized such that any existing
+  ;   references to the string will internally see the new buffer and not the
+  ;   tree of chunks.
+  ; * All chunks in the tree (rope) are non-empty.
+  ; * Getting the length of the string in code units is always constant-time.
+  ; * If the string is a `Concat` (after concatenation) then slicing is
+  ;   constant-time if the offsets are reachable in the first or last slice of
+  ;   the tree.
+  ; 
+  ; Rough sketch of the type:
+  ; 
+  ;   type pstring
+  ;     = Slice { buffer :: pstring-buffer, offset :: fixnum, length :: fixnum }
+  ;     | Concat { length :: fixnum, str :: MutableRef (ConcatTree | Slice) }
+  ;
+  (define-record pstring ())
+  (define-record Slice pstring
+                 ((immutable buffer)
+                  (immutable offset)
+                  (immutable length)))
+  (define-record Concat pstring
+                 ((immutable length)
+                  (mutable str)))
+
+  (define (pstring-length str)
+    (cond
+      [(Slice? str) (Slice-length str)]
+      [else (Concat-length str)]))
+
+  ; A concatenation of two slices with constant-time access to the first and last slice
+  (define-record ConcatTree
+                 ((immutable prefix)
+                  (immutable deep)
+                  (immutable suffix)))
+
+  ; Slice -> ConcatTree -> ConcatTree
+  (define (tree-cons a b)
+    (make-ConcatTree
+      a
+      (if (null? (ConcatTree-deep b))
+        (ConcatTree-prefix b)
+        (cons (ConcatTree-prefix b) (ConcatTree-deep b)))
+      (ConcatTree-suffix b)))
+
+  ; ConcatTree -> Slice -> ConcatTree
+  (define (tree-snoc a b)
+    (make-ConcatTree
+      (ConcatTree-prefix a)
+      (if (null? (ConcatTree-deep a))
+        (ConcatTree-suffix a)
+        (cons (ConcatTree-deep a) (ConcatTree-suffix a)))
+      b))
+
+  ; ConcatTree -> ConcatTree -> ConcatTree
+  (define (concat a b)
+    (make-ConcatTree
+      (ConcatTree-prefix a)
+      (cond
+        [(and (null? (ConcatTree-deep a)) (null? (ConcatTree-deep b)))
+          (cons (ConcatTree-suffix a) (ConcatTree-prefix b))]
+        [(null? (ConcatTree-deep a))
+          (cons (cons (ConcatTree-suffix a) (ConcatTree-prefix b))
+                (ConcatTree-deep b))]
+        [(null? (ConcatTree-deep b))
+          (cons (ConcatTree-deep a)
+                (cons (ConcatTree-suffix a) (ConcatTree-prefix b)))]
+        [else
+          (cons
+            (cons (ConcatTree-deep a)
+                  (cons (ConcatTree-suffix a) (ConcatTree-prefix b)))
+            (ConcatTree-deep b))])
+      (ConcatTree-suffix b)))
+
+  (define (pstring-compact? str)
+    (and (Concat? str) (Slice? (Concat-str str))))
+
+  ; fixnum -> ConcatTree -> Slice
+  (define (compact-tree len str)
+    (define (copy-tree! bv at kons)
```

Comment:
This function could do with a docstring or more comments in the body, or longer variable names - I can't figure out what `bv` and `at` are..

Reply:
Added comments and renamed vars.